django/db/models/constraints.py

```diff
@@ -444,14 +444,20 @@ def validate(self, model, instance, exclude=None, using=DEFAULT_DB_ALIAS):
                         self.get_violation_error_message(),
                         code=self.violation_error_code,
                     )
-                # When fields are defined, use the unique_error_message() for
-                # backward compatibility.
-                for model, constraints in instance.get_constraints():
-                    for constraint in constraints:
-                        if constraint is self:
-                            raise ValidationError(
-                                instance.unique_error_message(model, self.fields),
-                            )
+                if self.violation_error_message == self.default_violation_error_message:
+                    # When fields are defined, use the unique_error_message() as
+                    # a default for backward compatibility.
+                    validation_error_message = instance.unique_error_message(
+                        model, self.fields
+                    )
+                    violation_error_code = None
```

Comment:
```suggestion
                    violation_error_code = validation_error_message.code
```
I think this might be more accurate and if I make this update, no tests fail (so we are missing some asserts here)